docs/architecture/adr-024-coin-metadata.md

```diff
@@ -0,0 +1,145 @@
+# ADR 024: Coin Metadata
+
+## Changelog
+
+- 05/19/2020: Initial draft
+
+## Status
+
+Proposed
+
+## Context
+
+Assets in the Cosmos SDK are represented via a `Coins` type that consists of an `amount` and a `denom`,
+where the `amount` can be any arbitrarily large or small value. In addition, the Cosmos SDK uses an
+account-based model where there are two types of primary accounts -- basic accounts and module accounts.
+All account types have a set of balances that are composed of `Coins`. The `x/bank` module keeps
+track of all balances for all accounts and also keeps track of the total supply of balances in an
+application.
+
+With regards to a balance `amount`, the Cosmos SDK assumes a static and fixed unit of denomination,
+regardless of the denomination itself. In other words, clients and apps built atop a Cosmos-SDK-based
+chain may choose to define and use arbitrary units of denomination to provide a richer UX, however, by
```

Comment:
What about changing bank and supply to use sdk.Dec
Then there is no confusion of units.

Reply:
That is much more radical of a change. I personally prefer the model of just using integers (same as ethereum).

Reply:
I would like too see support for Dec coins in the future but don't think it needs to come in this PR. I do think there is a clean way to do it that doesn't require all interfaces to support decimals and have been intending to write down my thoughts in a future ADR. I do think this denom metadata is a good basis for future Dec support

Reply:
It is much easier working with integers in the framework and I'd rather keep it this way, but ultimately dev input should drive this decision and so I would love to see input from other stakeholders and SDK users.

This topic can easily lead to bikeshedding so let's raise the concern in another issue and so I'm going to resolve this comment.

Reply:
The primary reason for sticking with using Integers instead of Decimals is that most client library languages don't have a native type similar to sdk.Dec, as in most languages, floats are non-deterministic and are binary-based.  If coin amounts are sdk.Dec, then we would need to recreate the "sdk.Dec" type in all the client libraries/languages.

(For uses of sdk.Dec in other modules, we can usually get away with using them as Strings or BigNum on client side, but for something like coin amounts, that basically every single wallet or client will want to support and be able to do arithmetic on, basic integers is the most universal)

Reply:
So I've ran into this trying to design some custom modules for Regen. So consider me still an SDK user.

I understand the client lib issue, but that's not a problem if you support both Int and Dec as two parallel APIs (with fractional balances just getting truncated for Int only API users).

Anyway, as I said I don't think it's an issue to address for this PR so let's pause till a separate discussion if it's still relevant.